server/logging/kinesis.go

```diff
@@ -0,0 +1,204 @@
+package logging
+
+import (
+	"context"
+	"encoding/json"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/credentials/stscreds"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/aws/aws-sdk-go/service/kinesis/kinesisiface"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+)
+
+const (
+	kinesisMaxRetries = 8
+
+	// See
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/kinesis/#Kinesis.PutRecords
+	// for documentation on limits.
+	kinesisMaxRecordsInBatch = 500
+	kinesisMaxSizeOfRecord   = 1000 * 1000     // 1,000 KB
+	kinesisMaxSizeOfBatch    = 5 * 1000 * 1000 // 5 MB
+)
+
+type kinesisLogWriter struct {
+	client kinesisiface.KinesisAPI
+	stream string
+	logger log.Logger
+	rand   *rand.Rand
+}
+
+func NewKinesisLogWriter(region, id, secret, stsAssumeRoleArn, stream string, logger log.Logger) (*kinesisLogWriter, error) {
+	conf := &aws.Config{
+		Region: &region,
+	}
+
+	// Only provide static credentials if we have them
+	// otherwise use the default credentials provider chain
+	if id != "" && secret != "" {
+		conf.Credentials = credentials.NewStaticCredentials(id, secret, "")
+	}
+
+	sess, err := session.NewSession(conf)
+	if err != nil {
+		return nil, errors.Wrap(err, "create Kinesis client")
+	}
+
+	if stsAssumeRoleArn != "" {
+		creds := stscreds.NewCredentials(sess, stsAssumeRoleArn)
+		conf.Credentials = creds
+
+		sess, err = session.NewSession(conf)
+
+		if err != nil {
+			return nil, errors.Wrap(err, "create Kinesis client")
+		}
+	}
+	client := kinesis.New(sess)
+
+	// This will be used to generate random partition keys to balance
+	// records across Kinesis shards.
+	rand := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	k := &kinesisLogWriter{
+		client: client,
+		stream: stream,
+		logger: logger,
+		rand:   rand,
+	}
+	if err := k.validateStream(); err != nil {
+		return nil, errors.Wrap(err, "create Kinesis writer")
+	}
+	return k, nil
+}
+
+func (k *kinesisLogWriter) validateStream() error {
+	out, err := k.client.DescribeStream(
+		&kinesis.DescribeStreamInput{
+			StreamName: &k.stream,
+		},
+	)
+	if err != nil {
+		return errors.Wrapf(err, "describe stream %s", k.stream)
+	}
+
+	if (*(*out.StreamDescription).StreamStatus) != kinesis.StreamStatusActive {
```

Comment:
Do we need nil checks here? It's not immediately clear to me that this are guaranteed to be non-nil.

Reply:
AWS's API says that it should return StreamStatus with this type of request. It's marked as 'required' in the Go code. If we don't get a non-nil error, I don't think it's possible to have these be nil.

We could add them, but I don't think it's needed. (aside: This follows the same pattern as seen in the existing Firehose logger.)